syntax = "proto3";

package TW.Aion.Proto;
option java_package = "wallet.core.jni.proto";

// Transaction payload (transfer, smart contract call, ...)
message Payload {
    // Native coin transfer transaction
    message Transfer {
        // Amount to send in wei (256-bit number)
        bytes amount = 1;

        // Optional payload
        bytes payload = 2;
    }

    // Generic smart contract transaction
    message ContractGeneric {
        // Optional payload
        bytes payload = 1;
    }

    oneof payload_oneof {
        Transfer payload_transfer = 1;
        ContractGeneric payload_contract_generic = 4;
    }
}

// Input data necessary to create a signed transaction.
message SigningInput {
    // Nonce (256-bit number)
    bytes nonce = 1;

    // Gas price (256-bit number)
    bytes gas_price = 2;

    // Gas limit (256-bit number)
    bytes gas_limit = 3;

    // Recipient's address.
    string to_address = 4;

    // Private key.
    bytes private_key = 5;

    // Timestamp
    uint64 timestamp = 6;

    Payload payload = 7;
}

// Transaction signing output.
message SigningOutput {
    // Signed and encoded transaction bytes.
    bytes encoded = 1;

    // Signature.
    bytes signature = 2;
}
